### PR TITLE
PP-1776 Provider level gateway clients

### PR DESCRIPTION
**Pull request by imranbohoran**
## WHAT
_A brief description of the pull request:_

- As part of allowing for retry on captures, we have decided to configure our gateway clients to timeout earlier than wait for the egress proxy to timeout. However this also means that we want to have the gateway clients to be configurable per operation(i.e. auth, capture, client, refund).
- This PR provides the capability to configure clients for each gateway and for each operation.
- The timeout for capture in particular has been reduced to 10s as there's no point on us waiting for 60s for it to timeout (based on the metrics we have for capture response times). It either succeeds in about 600ms or it doesn't for over 1 minute.

## Still to do

The newly introduced [GatewayOperation](https://github.com/alphagov/pay-connector/blob/PP-1776_provider_level_gateway_clients/src/main/java/uk/gov/pay/connector/service/GatewayOperation.java) has some overlap in purpose with [PaymentGatewayName](https://github.com/alphagov/pay-connector/blob/PP-1776_provider_level_gateway_clients/src/main/java/uk/gov/pay/connector/resources/PaymentGatewayName.java). We think we should refactor to remove this duplication and intend to do that in a subsequent pull request. However, we're keen to ship this code increment as-is and do the refactoring in another PR.

with @heathd

### CODE REVIEW

**Comment by ghost** [comment]
There were no errors, go have a cup of coffee...

File: src/main/java/uk/gov/pay/connector/service/BasePaymentProvider.java [hunk 1]
```diff
@@ -35,19 +36,36 @@ public BasePaymentProvider(GatewayClient client, boolean isNotificationEndpointS
...
                         .postRequestFor(target, request.getGatewayAccount(), order.apply(request))
                         .bimap(
                                 GatewayResponse::with,
-                                r -> mapToResponse(r, clazz, responseIdentifier)
+                                r -> mapToResponse(r, clazz, responseIdentifier, gatewayClient)
+                        )
+        );
+    }
+
+    protected <U extends GatewayRequest> GatewayResponse sendReceive(U request, GatewayOperation operation, Function<U, GatewayOrder> order,
```
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/BasePaymentProvider.java]: This method is currently not been used - that's perfectly fine as it is here in preparation for the rest of the refactoring. However this new method is practically a copy-paste of the one up above which seems like a code smell. Should we not try to refactor the two methods to make one call the other as it is typically done with overloads?

File: src/main/java/uk/gov/pay/connector/service/ClientFactory.java [hunk 2]
```diff
@@ -52,11 +48,28 @@ public Client createWithDropwizardClient(String name) {
...
+        OperationOverrides overrides = getOverridesFor(operation, gateway);
+        if (overrides != null && overrides.getReadTimeout() != null) {
+            return (int) overrides.getReadTimeout().toMilliseconds();
+        }
+        return (int) conf.getCustomJerseyClient().getReadTimeout().toMilliseconds();
+    }
+
+    private OperationOverrides getOverridesFor(GatewayOperation operation, SupportedPaymentGateway gateway) {
+
+        if (conf.getGatewayConfigFor(gateway).getJerseyClientOverrides() != null) {
```
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/ClientFactory.java]: `conf.getGatewayConfigFor(gateway).getJerseyClientOverrides` could be extracted into a local variable to improve readability.

File: src/main/java/uk/gov/pay/connector/service/PaymentProviders.java [hunk 3]
```diff
@@ -30,36 +32,56 @@
...
+    private GatewayClient gatewayClientForOperation(SupportedPaymentGateway gateway,
+                                                    GatewayOperation operation,
+                                                    BiFunction<GatewayOrder, Invocation.Builder, Invocation.Builder> sessionIdentifier) {
+        return gatewayClientFactory.createGatewayClient(
+                gateway, operation, config.getGatewayConfigFor(gateway).getUrls(),
+                MediaType.APPLICATION_XML_TYPE, sessionIdentifier,
+                environment.metrics());
+    }
+
+    private PaymentProvider createSmartPayProvider(GatewayConfig config,
```
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/PaymentProviders.java]: `config` parameter is not needed anymore - may be a good opportunity to remove it.

File: src/main/resources/config/config.yaml [hunk 4]
```diff
@@ -40,7 +49,12 @@ transactionsPaginationServiceConfig:
   displayPageSize: ${TRANSACTION_LIST_DISPLAY_SIZE:-500}
 
 jerseyClient:
+  # Defines the socket timeout (SO_TIMEOUT), which is the
```
Comment (ckalista) [inline src/main/resources/config/config.yaml]: 👍  for the documentation.

**Review by ckalista — CHANGES_REQUESTED** [review]
Code review completed - looks good to me 👍

File: src/main/java/uk/gov/pay/connector/service/GatewayOperationClientBuilder.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
+package uk.gov.pay.connector.service;
+
+import java.util.EnumMap;
+
+public class GatewayOperationClientBuilder {
+
+    EnumMap<GatewayOperation, GatewayClient> operationGatewayClientEnumMap;
+
+    public GatewayOperationClientBuilder() {
```
Comment (daniele-occhipinti) [inline src/main/java/uk/gov/pay/connector/service/GatewayOperationClientBuilder.java]: It seems this constructor can be made private.

File: src/main/resources/config/config.yaml [hunk 6]
```diff
@@ -25,6 +25,15 @@ worldpay:
...
   credentials: ['username','password','merchant_id']
+  jerseyClientOverrides:
+    auth:
+      readTimeout: 90000ms
+    cancel:
+      readTimeout: 90000ms
+    refund:
+      readTimeout: 90000ms
+    capture:
+      readTimeout: 10000ms
```
Comment (daniele-occhipinti) [inline src/main/resources/config/config.yaml]: Maybe expressing those timeouts in seconds is more readable (fewer zeros). I am not sure whether that is supported. On the other hand, I understand timeouts are usually expressed in ms.

File: src/test/java/uk/gov/pay/connector/service/smartpay/SmartpayPaymentProviderTest.java [hunk 7]
```diff
@@ -44,37 +49,67 @@
 import static org.hamcrest.core.IsSame.sameInstance;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
-import static org.mockito.Matchers.anyObject;
-import static org.mockito.Matchers.anyString;
-import static org.mockito.Matchers.eq;
+import static org.mockito.Matchers.*;
+import static org.mockito.Matchers.any;
```
Comment (daniele-occhipinti) [inline src/test/java/uk/gov/pay/connector/service/smartpay/SmartpayPaymentProviderTest.java]: is this line necessary, given the line above?

**Review by daniele-occhipinti — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
There were errors, for info, please see...

**Comment by ghost** [comment]
There were errors, for info, please see...

File: src/main/java/uk/gov/pay/connector/service/BasePaymentProvider.java (hunk 1 above)
Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/service/BasePaymentProvider.java]: 👍 It was a hangover and unnecessary override from the previous implementation. Removed it.

File: src/main/java/uk/gov/pay/connector/service/ClientFactory.java (hunk 2 above)
Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/service/ClientFactory.java]: 👍 Done

File: src/main/java/uk/gov/pay/connector/service/PaymentProviders.java (hunk 3 above)
Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/service/PaymentProviders.java]: 👍 Yes, its not need indeed. Removed.

File: src/main/resources/config/config.yaml (hunk 4 above)
Comment (imranbohoran) [reply src/main/resources/config/config.yaml]: I'll 👍 your 👍 , for reading it 😄

File: src/main/resources/config/config.yaml (hunk 6 above)
Comment (imranbohoran) [reply src/main/resources/config/config.yaml]: This is going to change and will be less number of millis, so not changing anything right now.

File: src/main/java/uk/gov/pay/connector/service/GatewayOperationClientBuilder.java (hunk 5 above)
Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/service/GatewayOperationClientBuilder.java]: 👍 It certainly could. Done.

File: src/test/java/uk/gov/pay/connector/service/smartpay/SmartpayPaymentProviderTest.java (hunk 7 above)
Comment (imranbohoran) [reply src/test/java/uk/gov/pay/connector/service/smartpay/SmartpayPaymentProviderTest.java]: Yes it's needed. If not there are conflicts with `any` from another library.

**Comment by ghost** [comment]
There were no errors, go have a cup of coffee...

**Review by daniele-occhipinti — APPROVED** [review]

**Review by ckalista — APPROVED** [review]

**Pull request merged**